unit_testing/coding-challange-yaml.test.js

```diff
@@ -0,0 +1,202 @@
+      console.log("Here is the JSON:");
+      console.log(indentedJson);
+    }
+  });
+
+  // Checks that the YAML page has the required properties. Does this by
+  // oppening yaml, then using the "hasOwnProperty" method to check if it has
+  // all the required properties.
+  //    The required properties are: "title", "video_number", "date", "video_id"
+  // these properties were found at the Content Contribution Guide:
+  // https://github.com/CodingTrain/website/wiki/Content-Contribution-Guide
+  // The required properties for the index.md page was set to be the properties
+  // that apperead across all index.md pages in all directories.
+  test("Check that YAML contents of "+CodingChallangesDir+"/"+file
+  +" have required properties", () => {
+    let contents = fs.readFileSync(CodingChallangesDir+"/"+file, 'utf8');
+    let yamlContents = contents.split("---")[1];
+    const frontYaml = yaml.safeLoad(yamlContents);
+    let requiredProperties = ["title", "video_number", "date", "video_id"];
+    if (file=="index.md") {
+      requiredProperties = ["title", "layout"];
+    }
+    for (let key of requiredProperties) {
+      expect(frontYaml.hasOwnProperty(key)).toBe(true);
+    }
+  });
+
+  // Checks if object created from YAML has any undefined properties using
+  // checkHasUndefinedReccursive function (defined line 32).
+  //     I didn't cause an error to website when videos property was null. I
+  // don't know if undefined properties will always be handled cleanly. if so,
+  // maybe add a "strict" variable, and only run this test if trict is set to
+  // true.
+  test("Check no undefined properties of object from YAML of "+
+    CodingChallangesDir+"/"+file, () => {
+    let contents = fs.readFileSync(CodingChallangesDir+"/"+file, 'utf8');
+    let yamlContents = contents.split("---")[1];
+    const frontYaml = yaml.safeLoad(yamlContents);
+    if (checkHasUndefinedReccursive(frontYaml)) {
+      console.log("There was an undefined property in the object created from the YAML of "+
+        CodingChallangesDir+"/"+file+" here is the object:");
+      console.log(frontYaml);
+    }
+    expect(checkHasUndefinedReccursive(frontYaml)).toBe(false);
+  });
+
+  // Checks that all contributions have the required properties using
+  // .hasOwnProperty method.
+  test("Check that contributions of "+ CodingChallangesDir+"/"+file
+  +" has required properties", () => {
+    let contents = fs.readFileSync(CodingChallangesDir+"/"+file, 'utf8');
+    let yamlContents = contents.split("---")[1];
+    const frontYaml = yaml.safeLoad(yamlContents);
+    if (frontYaml.hasOwnProperty("contributions")) {
+      for (let contibutor of frontYaml["contributions"]) {
+        expect(contibutor.hasOwnProperty("title")).toBe(true);
+        expect(contibutor.hasOwnProperty("url")
+        || contibutor.hasOwnProperty("source")).toBe(true);
+        /************ IMPORTANT NOTE **********************
```

Comment:
I think this is the right choice, but perhaps @nilsweb @shiffman can review that we would like to have at least one of these and we can make a change to the contributing guide.

Reply:
Sounds like a good plan to me!